File: src/calculus_implicit_differentiation_easy.py
"""
calculus - implicit_differentiation (easy)
Generated: 2026-02-22T05:04:36.786935
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 6)

    if problem_type == 1:
        # ax^2 + by^2 = c (ellipse/hyperbola)
        a_val = nonzero(-6, 6)
        b_val = nonzero(-6, 6)
        c_val = randint(1, 30)
        # 2a x + 2b y dy/dx = 0 => dy/dx = -ax/(by)
        ans = simplify(Rational(-a_val, b_val) * x / y)
        return problem(
            question=f"Find $\\frac{{dy}}{{dx}}$ for ${latex(a_val*x**2 + b_val*y**2)} = {c_val}$.",
            answer=ans,
            difficulty=(1000, 1150),
            topic="calculus/implicit_differentiation",
            solution=steps(
                f"Differentiate: ${latex(2*a_val*x)} + {latex(2*b_val*y)}\\frac{{dy}}{{dx}} = 0$",
                f"$\\frac{{dy}}{{dx}} = {latex(ans)}$"
            ),
        )

    elif problem_type == 2:
        # Linear: ax + by = c, dy/dx = -a/b (constant)
        a_val = nonzero(-6, 6)
        b_val = nonzero(-6, 6)
        c_val = randint(-15, 15)
        ans = Rational(-a_val, b_val)
        return problem(
            question=f"Find $\\frac{{dy}}{{dx}}$ for ${latex(a_val*x + b_val*y)} = {c_val}$.",
            answer=ans,
            difficulty=(1000, 1100),
            topic="calculus/implicit_differentiation",
            solution=steps(
                f"Differentiate: ${a_val} + {b_val}\\frac{{dy}}{{dx}} = 0$",
                f"$\\frac{{dy}}{{dx}} = {latex(ans)}$"
            ),
        )

    elif problem_type == 3:
        # xy = k
        k_val = randint(2, 30)
        ans = -y/x
        return problem(
            question=f"Find $\\frac{{dy}}{{dx}}$ for $xy = {k_val}$.",
            answer=ans,
            difficulty=(1050, 1200),
            topic="calculus/implicit_differentiation",
            solution=steps(
                f"Product rule: $y + x\\frac{{dy}}{{dx}} = 0$",
                f"$\\frac{{dy}}{{dx}} = -\\frac{{y}}{{x}}$"
            ),
        )

    elif problem_type == 4:
        # x^2 + y^2 = r^2 (circle)
        r_sq = randint(4, 50)
        ans = -x/y
        return problem(
            question=f"Find $\\frac{{dy}}{{dx}}$ for $x^2 + y^2 = {r_sq}$.",
            answer=ans,
            difficulty=(1000, 1100),
            topic="calculus/implicit_differentiation",
            solution=steps(
                f"Differentiate: $2x + 2y\\frac{{dy}}{{dx}} = 0$",
                f"$\\frac{{dy}}{{dx}} = -\\frac{{x}}{{y}}$"
            ),
        )

    elif problem_type == 5:
        # ax^n + by = c, dy/dx = -an*x^(n-1)/b
        a_val = nonzero(-5, 5)
        n_val = randint(2, 4)
        b_val = nonzero(-5, 5)
        c_val = randint(1, 20)
        # a*n*x^(n-1) + b*dy/dx = 0
        ans = simplify(Rational(-a_val * n_val, b_val) * x**(n_val - 1))
        return problem(
            question=f"Find $\\frac{{dy}}{{dx}}$ for ${latex(a_val*x**n_val + b_val*y)} = {c_val}$.",
            answer=ans,
            difficulty=(1100, 1250),
            topic="calculus/implicit_differentiation",
            solution=steps(
                f"Differentiate: ${latex(a_val*n_val*x**(n_val-1))} + {b_val}\\frac{{dy}}{{dx}} = 0$",
                f"$\\frac{{dy}}{{dx}} = {latex(ans)}$"
            ),
        )

    else:
        # ax*y^2 = c, differentiate
        a_val = nonzero(-4, 4)
        c_val = randint(1, 20)
        # a*y^2 + a*x*2y*dy/dx = 0 => dy/dx = -y/(2x)
        ans = -y / (2 * x)
        return problem(
            question=f"Find $\\frac{{dy}}{{dx}}$ for ${latex(a_val*x*y**2)} = {c_val}$.",
            answer=ans,
            difficulty=(1100, 1250),
            topic="calculus/implicit_differentiation",
            solution=steps(
                f"Product rule: ${a_val}y^2 + {a_val}x \\cdot 2y\\frac{{dy}}{{dx}} = 0$",
                f"$\\frac{{dy}}{{dx}} = -\\frac{{y}}{{2x}}$"
            ),
        )

emit(generate())

File: src/calculus_implicit_differentiation_easy.jl
# calculus - implicit_differentiation (easy)
# Generated: 2026-03-08T20:40:15.088048

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("calculus/implicit_differentiation")
    
    problem_type = rand(1:4)
    
    if problem_type == 1
        # Circle: x^2 + y^2 = r^2, find dy/dx
        r_sq = randint(4, 100)
        lhs = x^2 + y^2
        
        # dy/dx = -x/y
        dydx = -x/y
        
        problem(
            question="Find \\frac{dy}{dx} using implicit differentiation: \$$(tex(lhs)) = $(r_sq)\$",
            answer=dydx,
            difficulty=(700, 900),
            solution=steps(
                sol("Given", lhs ~ r_sq),
                "Differentiate both sides with respect to x: \$2x + 2y\\frac{dy}{dx} = 0\$",
                "Solve for \\frac{dy}{dx}: \$2y\\frac{dy}{dx} = -2x\$",
                sol("Answer", dydx)
            ),
            time=90
        )
        
    elseif problem_type == 2
        # Linear combination: ax + by = c, find dy/dx
        a = nonzero(-8, 8)
        b = nonzero(-8, 8)
        c = randint(-30, 30)
        lhs = a*x + b*y
        
        # dy/dx = -a/b
        dydx = -a//b
        
        problem(
            question="Find \\frac{dy}{dx} using implicit differentiation: \$$(tex(lhs)) = $(c)\$",
            answer=dydx,
            difficulty=(700, 850),
            solution=steps(
                sol("Given", lhs ~ c),
                "Differentiate both sides: \$$(a) + $(b)\\frac{dy}{dx} = 0\$",
                "Solve for \\frac{dy}{dx}: \$$(b)\\frac{dy}{dx} = $(-a)\$",
                sol("Answer", dydx)
            ),
            time=75
        )
        
    elseif problem_type == 3
        # x^2 + y^2 = axy type, find dy/dx
        a = nonzero(-6, 6)
        lhs = x^2 + y^2
        rhs = a*x*y
        
        # Differentiate: 2x + 2y(dy/dx) = a*y + a*x(dy/dx)
        # dy/dx = (ay - 2x)/(2y - ax)
        dydx_num = a*y - 2*x
        dydx_den = 2*y - a*x
        dydx = dydx_num / dydx_den
        
        problem(
            question="Find \\frac{dy}{dx} using implicit differentiation: \$$(tex(lhs)) = $(tex(rhs))\$",
            answer=dydx,
            difficulty=(900, 1100),
            solution=steps(
                sol("Given", lhs ~ rhs),
                "Differentiate both sides: \$2x + 2y\\frac{dy}{dx} = $(a)y + $(a)x\\frac{dy}{dx}\$",
                "Collect terms with \\frac{dy}{dx}: \$(2y - $(a)x)\\frac{dy}{dx} = $(a)y - 2x\$",
                sol("Answer", dydx)
            ),
            time=120
        )
        
    else
        # x^2 + ay^2 = c, find dy/dx
        a = nonzero(-7, 7)
        c = randint(10, 80)
        lhs = x^2 + a*y^2
        
        # dy/dx = -x/(ay)
        dydx = -x/(a*y)
        
        problem(
            question="Find \\frac{dy}{dx} using implicit differentiation: \$$(tex(lhs)) = $(c)\$",
            answer=dydx,
            difficulty=(750, 950),
            solution=steps(
                sol("Given", lhs ~ c),
                "Differentiate both sides: \$2x + $(2*a)y\\frac{dy}{dx} = 0\$",
                "Solve for \\frac{dy}{dx}: \$$(2*a)y\\frac{dy}{dx} = -2x\$",
                sol("Answer", dydx)
            ),
            time=90
        )
    end
end